beanie/odm/fields.py

```diff
@@ -147,17 +148,36 @@ def _validate(cls, v):
         def __get_pydantic_core_schema__(
             cls, source_type: Type[Any], handler: GetCoreSchemaHandler
         ) -> CoreSchema:
+            if not IS_PYDANTIC_V2_10:
+                return json_or_python_schema(
+                    python_schema=no_info_plain_validator_function(
+                        cls._validate
+                    ),
+                    json_schema=no_info_plain_validator_function(
+                        cls._validate,
+                        metadata={
+                            "pydantic_js_input_core_schema": str_schema(
+                                pattern="^[0-9a-f]{24}$",
+                                min_length=24,
+                                max_length=24,
+                            )
+                        },
+                    ),
+                    serialization=plain_serializer_function_ser_schema(
+                        lambda instance: str(instance),
+                        return_schema=str_schema(),
+                        when_used="json",
+                    ),
+                )
             return json_or_python_schema(
```

Comment:
Hello, Pydantic maintainer here.

I don't think there's a need to make a distinction between JSON and Python here. `pydantic_js_input_core_schema` can be specified for both, it's not an issue:

```python
return no_info_plain_validator_function(
    cls._validate,
    json_schema_input_schema=...,
    serialization=...,
)
```

As a side note, I know our logic to create custom types through `__get_pydantic_core_schema__` isn't ideal (confusing pattern, need to create core schemas manually, etc). Ideally, `PydanticObjectId` could be defined as an annotated type:

```python
PydanticObjectId = Annotated[
    ObjectId,
    PlainValidator(_validate, json_schema_input_type=Annotated[str, Field(pattern="^[0-9a-f]{24}$", min_length=24, max_length=24)]),
    PlainSerializer(lambda v: str(v), return_type=str),
]
```

This way, you go through the "user facing" validators API, where `json_schema_input_type` is guaranteed to not change (unlike _how_ this attribute was stored on the core schema — we had to change it to avoid core schema simplification errors).

Reply:
Thanks for the suggestion. I updated on the plain validator function. But for the 

```python3
PydanticObjectId = Annotated[
    ObjectId,
    PlainValidator(_validate, json_schema_input_type=Annotated[str, Field(pattern="^[0-9a-f]{24}$", min_length=24, max_length=24)]),
    PlainSerializer(lambda v: str(v), return_type=str),
]
```

I have noticed there are some breaking changes that would require significant modifications to the codebase since it changes the checker for `PydanticObjectId` as `typing.Annotated` and not a normal class. One of the example is:

```python3
def extract_id_class(annotation) -> Type[Any]:
    if get_origin(annotation) is not None:
        try:
            annotation = next(
                arg for arg in get_args(annotation) if arg is not type(None)
            )
        except StopIteration:
            annotation = None
    if inspect.isclass(annotation):
        return annotation
    raise ValueError("Unknown annotation: {}".format(annotation))
```

Reply:
Yes indeed this will most likely break things (and the `Annotated` form also has some drawbacks). And it seems that beanie expects `PydanticObjectId` to be instantiable, which is not going to be the case with `Annotated` :/